wazo_call_logd/generator.py

```diff
@@ -91,57 +91,84 @@ def _remove_duplicate_participants(self, call_log):
                 continue
-            raw_attributes.update(**confd_participant)
-            answered = {}
+            raw_attributes.update(**confd_participant._asdict())
+
+            participant = CallLogParticipant(user_uuid=confd_participant.uuid)
+            participant.line_id = confd_participant.line_id
+            participant.tags = confd_participant.tags
+            participant.role = raw_attributes['role']
             if 'answered' in raw_attributes:
-                answered['answered'] = raw_attributes['answered']
-            participant_model = CallLogParticipant(
-                user_uuid=confd_participant['uuid'],
-                line_id=confd_participant['line_id'],
-                tags=confd_participant['tags'],
-                role=raw_attributes['role'],
-                **answered,
-            )
-            call_log.participants.append(participant_model)
-
-    def _update_call_participants_with_their_tags(self, confd, call_log):
-        source_user_uuid = call_log.source_user_uuid
-        destination_user_uuid = call_log.destination_user_uuid
-
-        if source_user_uuid:
-            source_user = confd.users.get(source_user_uuid)
-            source_user_tags = (
-                [tag.strip() for tag in source_user['userfield'].split(',')]
-                if source_user['userfield']
-                else []
-            )
+                participant.answered = raw_attributes["answered"]
+            connected_participants.append(participant)
+        return connected_participants
 
-        if destination_user_uuid:
-            destination_user = confd.users.get(destination_user_uuid)
-            destination_user_tags = (
-                [tag.strip() for tag in destination_user['userfield'].split(',')]
-                if destination_user['userfield']
-                else []
-            )
+    def _fetch_participants(self, confd: ConfdClient, call_log: RawCallLog):
+        users: dict[str, ParticipantInfo] = {}
+
+        def get_user(user_uuid):
+            confd_participant = users.get(user_uuid)
+            if not confd_participant:
+                users[user_uuid] = confd_participant = find_participant_by_uuid(
+                    confd, user_uuid
+                )
+            return confd_participant
+
+        connected_participants = self._compute_participants_from_channels(confd, call_log)
+
+        # participant information from CEL interpretation can augment participants extracted from channels
+        # and fill in unreachable participants with no corresponding channel
+        unreached_participants = []
+        users_from_cel = (
+            str(participant_info["user_uuid"])
+            for participant_info in call_log.participants_info
+            if "user_uuid" in participant_info
+        )
+        users_from_channels = (str(p.user_uuid) for p in connected_participants)
+        for user_uuid in uniq(chain(users_from_cel, users_from_channels)):
```

Comment:
Why not using `set`? Maybe not most efficient, but how much more easier to maintain/read than the custom `uniq` method
```suggestion
        users_from_cel = (
            str(participant_info["user_uuid"])
            for participant_info in call_log.participants_info
            if "user_uuid" in participant_info
        )
        users_from_channels = (str(p.user_uuid) for p in connected_participants)
        all_unique_users = set([*users_from_cel, *users_from_channels])
        for user_uuid in all_unique_users:
```

Reply:
I did start out using `set`s, but decided on this solution to maintain ordering of events.

Reply:
I went back to sets. This solution does not guarantee meaningful ordering anyway, I think we'd have to track order of appearance of events explicitly for that to be consistently possible.